library/common/http/dispatcher.cc

```diff
@@ -35,32 +36,41 @@ void Dispatcher::DirectStreamCallbacks::encodeHeaders(const ResponseHeaderMap& h
             direct_stream_.stream_handle_, end_stream, headers);
 
   ASSERT(http_dispatcher_.getStream(direct_stream_.stream_handle_));
-
-  uint64_t response_status = Utility::getResponseStatus(headers);
-  // Track success for later bookkeeping (stream could still be reset).
-  success_ = CodeUtility::is2xx(response_status);
-
   if (end_stream) {
     closeStream();
   }
 
-  // TODO: ***HACK*** currently Envoy sends local replies in cases where an error ought to be
-  // surfaced via the error path. There are ways we can clean up Envoy's local reply path to
-  // make this possible, but nothing expedient. For the immediate term this is our only real
-  // option. See https://github.com/lyft/envoy-mobile/issues/460
+  uint64_t response_status = Utility::getResponseStatus(headers);
+
+  // Presence of internal error header indicates an error that should be surfaced as an
+  // error callback (rather than an HTTP response).
+  const auto error_code_header = headers.get(InternalHeaders::get().ErrorCode);
+  if (!error_code_header.empty()) {
+    envoy_error_code_t error_code;
+    bool check = absl::SimpleAtoi(error_code_header[0]->value().getStringView(), &error_code);
+    RELEASE_ASSERT(check, "parse error reading error code");
+    error_code_ = error_code;
+
+    const auto error_message_header = headers.get(InternalHeaders::get().ErrorMessage);
+    if (!error_message_header.empty()) {
+      error_message_ =
+          Buffer::Utility::copyToBridgeData(error_message_header[0]->value().getStringView());
+    }
+
+    uint32_t attempt_count;
+    if (headers.EnvoyAttemptCount() &&
+        absl::SimpleAtoi(headers.EnvoyAttemptCount()->value().getStringView(), &attempt_count)) {
+      error_attempt_count_ = attempt_count;
+    }
 
-  // Error path: missing EnvoyUpstreamServiceTime implies this is a local reply, which we treat as
-  // a stream error.
-  if (!success_ && headers.get(Headers::get().EnvoyUpstreamServiceTime).empty()) {
-    ENVOY_LOG(debug, "[S{}] intercepted local response", direct_stream_.stream_handle_);
-    mapLocalResponseToError(headers);
     if (end_stream) {
       onError();
     }
     return;
   }
 
-  // Normal response path.
+  // Track success for later bookkeeping (stream could still be reset).
+  success_ = CodeUtility::is2xx(response_status);
```

Comment:
related to my comment above re: 218. Should the local error path be counted as success?

Reply:
It won't be. The function returns a couple lines above.